cql-pytest/test_select_from_mutation_fragments: do not share test_table

bhalevy opened the pull request:
Make the test_table fixture per function,
so each test can control autocompaction on its own.

denesb reviewed (approved):
LGTM, was there a problem with this test?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3927/

bhalevy commented:
> LGTM, was there a problem with this test?

Only in my branch, but it's nicer this way in any case.

denesb commented:
> > LGTM, was there a problem with this test?
> 
> Only in my branch, but it's nicer this way in any case.

Can you please add the problem to the patch description? Originally I explicitly wanted to have a shared table to reduce the number of table create/destroy. I'm fine with undoing this but I'm curious what the problem you saw was.

bhalevy commented:
I don't remember the exact problem since it was in an old version of https://github.com/scylladb/scylladb/pull/14043
For testing purposes I wanted to disable auto compaction in this cql test so to reduce the noise for row cache updates and so I needed the isolation between test cases so each can disable/reenable autocompaction independentally.

The last part is mentioned in the patch description, maybe lacking details though.

bhalevy commented:
Reproduced this failure:
```
_______________________________ test_ck_in_query _______________________________

cql = <cassandra.cluster.Session object at 0x7f2bec1f8690>
test_table = 'cql_test_1695745874994.cql_test_1695745874998'

    def test_ck_in_query(cql, test_table):
        pk1 = util.unique_key_int()
        pk2 = util.unique_key_int()
    
        cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
        cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 1, 1, 'vv')")
        nodetool.flush(cql, f"{test_table}")
        cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
    
        ks, _ = test_table.split(".")
    
        with nodetool.no_autocompaction_context(cql, ks):
            sources_res = list(cql.execute(f"SELECT * FROM MUTATION_FRAGMENTS({test_table}) WHERE pk1 = {pk1} AND pk2 = {pk2}"))
    
            sources = {r.mutation_source.split(":")[0]: r.mutation_source for r in sources_res}
>           assert len(sources) == 3
E           AssertionError: assert 2 == 3
E            +  where 2 = len({'memtable': 'memtable:0', 'sstable': 'sstable:/home/bhalevy/dev/scylla/testlog/dev/scylla-5/data/cql_test_1695745874994/cql_test_1695745874998-1c973d605c8a11ee8ca1d3b9c10297c3/me-3g9p_19w4_1i0g124znm5qy76dab-big-Data.db'})
```

nyh commented on test/cql-pytest/test_select_from_mutation_fragments.py:
```diff
@@ -92,23 +92,25 @@ def expect_sources(*expected_sources):
-    cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
-    expect_sources('memtable')
+    ks, _ = test_table.split(".")
+    with nodetool.no_autocompaction_context(cql, ks):
```
I suspect that the addition and movement of no_autocompaction_context is your real fix to whichever problem you saw, and the "function" scope wasn't necessary.

nyh commented:
> I don't remember the exact problem since it was in an old version of #14043 For testing purposes I wanted to disable auto compaction in this cql test so to reduce the noise for row cache updates and so I needed the isolation between test cases so each can disable/reenable autocompaction independentally.

These existing tests test what they test - please don't make them (without writing this anywhere) test other stuff or make it convenient for you to debug unrelated stuff. You can always make local modifications without submitting them as PRs :-) If you do want to add stuff like disabling autocompaction without need, you should at least add a comment saying that this is unneeded and helped you debug this-and-that.

In any case, none of this is relevant to the function-scope change. As I noted in my review,  the no_autocompaction_context context manager that @denesb added can be used to temporarily disable autocompaction and reenable it at the end of the test. It should work even if the same table is shared by all tests. If it doesn't work, then we have a bug somewhere and should fix it - not sweep it under the rug.


> 
> The last part is mentioned in the patch description, maybe lacking details though.

bhalevy commented:
@nyh are the test cases in cql-pytest running in parallel to each other?

denesb commented:
> @nyh are the test cases in cql-pytest running in parallel to each other?

Not yet, but they [will](https://github.com/scylladb/scylladb/pull/15238).

bhalevy commented:
@denesb I realized that the tests insert data into the table right before disabling auto-compaction.
That is sensitive to timing since compaction might kick-in right before the api call to disable it.
I'm not sure why, but it's hard to reproduce with disabling auto-compaction using the per-keyspace storage_service api.
But when I disable auto-compaction using the column_family per-table api this is reproduced even with a small --repeat count, and the following patch helps:
```diff
diff --git a/test/cql-pytest/test_select_from_mutation_fragments.py b/test/cql-pytest/test_select_from_mutation_fragments.py
index 93f77b2f99..a2506f6f04 100644
--- a/test/cql-pytest/test_select_from_mutation_fragments.py
+++ b/test/cql-pytest/test_select_from_mutation_fragments.py
@@ -382,14 +382,14 @@ def test_ck_in_query(cql, test_table):
     pk1 = util.unique_key_int()
     pk2 = util.unique_key_int()
 
-    cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
-    cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 1, 1, 'vv')")
-    nodetool.flush(cql, f"{test_table}")
-    cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
-
     ks, _ = test_table.split(".")
 
     with nodetool.no_autocompaction_context(cql, test_table):
+        cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
+        cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 1, 1, 'vv')")
+        nodetool.flush(cql, f"{test_table}")
+        cql.execute(f"INSERT INTO {test_table} (pk1, pk2, ck1, ck2, v) VALUES ({pk1}, {pk2}, 0, 0, 'vv')")
+
         sources_res = list(cql.execute(f"SELECT * FROM MUTATION_FRAGMENTS({test_table}) WHERE pk1 = {pk1} AND pk2 = {pk2}"))
 
         sources = {r.mutation_source.split(":")[0]: r.mutation_source for r in sources_res}
```

Why not just create the test_table with compaction disabled for the whole module?
Like this:
```diff
diff --git a/test/cql-pytest/test_select_from_mutation_fragments.py b/test/cql-pytest/test_select_from_mutation_fragments.py
index 93f77b2f99..9badabb311 100644
--- a/test/cql-pytest/test_select_from_mutation_fragments.py
+++ b/test/cql-pytest/test_select_from_mutation_fragments.py
@@ -22,7 +22,8 @@ import util
 @pytest.fixture(scope="module")
 def test_table(cql, test_keyspace):
     """ Prepares a table for the mutation dump tests to work with."""
-    with util.new_test_table(cql, test_keyspace, 'pk1 int, pk2 int, ck1 int, ck2 int, v text, s text static, PRIMARY KEY ((pk1, pk2), ck1, ck2)') as table:
+    with util.new_test_table(cql, test_keyspace, 'pk1 int, pk2 int, ck1 int, ck2 int, v text, s text static, PRIMARY KEY ((pk1, pk2), ck1, ck2)',
+                             " WITH compaction = {'class':'SizeTieredCompactionStrategy', 'enabled':'false'}") as table:
         yield table
 
 
```

denesb commented:
@bhalevy I can't remember why I didn't use that or `NullCompactionStrategy`. But I don't see a reason not to use it, feel free to send a patch.

bhalevy commented:
Closing the PR in favor of a new one using `NullCompactionStrategy`

nyh commented:
> > @nyh are the test cases in cql-pytest running in parallel to each other?
> 
> Not yet, but they [will](https://github.com/scylladb/scylladb/pull/15238).

Please correct me if I'm wrong, but the intention there is to run multiple Scylla instances in parallel, and run tests against **different** Scylla instances in parallel. So we don't need to support parallel tests. It's not pretty and I don't like it, but we already have several tests which temporarily change configuration or something like that, and those won't work properly if run in parallel against the same Scylla. When I write tests I usually try to avoid relying on being the only thing running against Scylla (e.g., tests for logs or metrics need to allow for the possibility that other workloads are running in parallel), but if we ever want to really support running the tests in parallel against a single instance of Scylla - we'll need to fix a few tests.
